Add Error Reporting Prompt

3nvi opened the pull request:
## Background

This PR adds the necessary UI pieces to allow the user to opt-in or out of error reporting.

Closes #121 
Closes #152 

## Changes

- Simplify Appsync schema for getting & updating an Organization
- Add an initial popup prompt, to allow them to select their error reporting preference
- Add the option to modify their preferences  under general settings & re-organize form
- Add a check to see whether Panther should send errors to sentry
- Persist error reporting preferences  in localStorage, in order to catch errors occurring prior to the request that fetches the user's preferences
  

## Testing

Manually tested that:
- Error doesn't get reported if the user hasn't yet selected a preference
- Error doesn't get reported if the  user has opted out
- Error **does** get reported if the user has opted in

## Screenshots

Initially the user will see this (@jacknagz can we verify that the wording is ok?)

<img width="1680" alt="Screen Shot 2020-02-13 at 7 14 31 PM" src="https://user-images.githubusercontent.com/10436045/74460655-e195b580-4e95-11ea-83dd-6fa152927d99.png">

And he can opt in/out through this

<img width="1676" alt="Screen Shot 2020-02-13 at 7 14 16 PM" src="https://user-images.githubusercontent.com/10436045/74460649-de9ac500-4e95-11ea-9d1c-746d9b520534.png">

austinbyers commented on api/graphql/schema.graphql:
```diff
@@ -251,16 +247,18 @@ enum AlertReportFrequencyEnum {
 
 type Organization {
-  id: String
+  id: String!
```
The orgID is a dummy constant that will probably go away before v1, can we just remove it from the frontend now?

3nvi replied:
Updated

austinbyers commented on web/src/components/forms/analytics-consent-form/error-reporting-section.tsx:
```diff
@@ -0,0 +1,41 @@
+        </InputElementLabel>
+        <Text color="grey300" size="medium">
+          Crashes and runtime exceptions will be <b>anonymously</b> reported, in an effort to make
+          the Panther team aware of the related issues
```
Maybe something more concise:

> Help improve Panther by submitting <b>anonymized</b> crashes and runtime exceptions

austinbyers commented on deployments/core/appsync.yml:
```diff
@@ -201,55 +201,6 @@ Resources:
         Endpoint: !Ref RemediationApi
 
-  ########## Functions ##########
-  GetOrganizationFunction:
```
Just for my understanding, how were we able to remove this? Don't we need the function so AppSync knows how to talk to the `organizations-api`?

3nvi replied:
we needed the pipeline (which used 2 functions) in order to :

1. Get  the `awsConfig` from the first function in order to know the org that the user belongs to
2. Forward it to the 2nd function, so that the `updateOrganization` knows which orgId to update

In a single-tenant world, we only have 1 org, so this was pointless

jacknagz commented:
awesome, @3nvi! here are my suggestions:

### Help Improve Panther!

Opt-in to occasionally provide diagnostic information for improving reliability.
**All information is anonymized.**

- [ ] Report Web Application Errors
	       Crashes and runtime exceptions

- [ ] Future other option goes here
	       Future other description

						[ Save ]

3nvi commented:
> I won't pretend I understand the frontend code, but I would say just based on the screenshots it would be nice to have a small section explaining what exactly is being reported. Browser version? Error code? A stack trace? The exact parameters of the request that caused the crash (which themselves may contain some sensitive information)? If there is not enough space in the popup, perhaps we could document it somewhere else.

@nhakmiller I updated the code to remove sensitive information (parameters, variables) from being logged. I also added a related doc to describe all the things we are logging.

Let me know if it looks ok to you

3nvi commented on web/src/components/forms/analytics-consent-form/error-reporting-section.tsx:
```diff
@@ -0,0 +1,53 @@
+            color="grey300"
+            size="medium"
+            is="a"
+            href={`${PANTHER_SCHEMA_DOCS_LINK}/privacy/error-reporting`}
```
@nhakmiller @austinbyers is this link correct if I created a doc under `/docs/gitbook/privacy/error-reporting/`?

nhakmiller replied:
Assuming `PANTHER_SCHEMA_DOCS_LINK` is just `docs.runpanther.io` you should be good.

austinbyers replied:
You want `docs.runpanther.io/privacy/error-reporting`, like Nick said. But the `privacy` path will only exist if you have it as a header in  docs/SUMMARY.md

3nvi replied:
gotcha. Perfect!

nhakmiller commented on docs/gitbook/privacy/error-reporting.md:
```diff
@@ -0,0 +1,15 @@
+- The version of your Browser, OS & Panther installation
+  > This helps us know how to properly replicate the issue.
+- The type of error, its related stack trace and the URL of the page in which it occurred.
+  > This helps us identify the piece of code associated with the issue you are experiencing, while gaining insight on the predecing function calls. **All sensitive parameters or variables are excluded**.
```
I think there is a typo here, it should be:

`insight on the preceding function calls.`

3nvi replied:
good catch

austinbyers commented on docs/gitbook/privacy/error-reporting.md:
```diff
@@ -0,0 +1,15 @@
+everytime there is a web application exception or crash:
+
+- Your Public IPv4
+  > This helps us identify whether the same or a different user is experiencing the error. This information is not used in any other way.
```
In most cases, the public IP will be the same for all users on the same team, since they're probably behind NAT. In the EU, [IP addresses can be considered PII](https://www.enterprisetimes.co.uk/2016/10/20/ecj-rules-ip-address-is-pii/)

Can we avoid collecting IPs? I definitely wouldn't call IP collection *anonymized* error reporting

3nvi replied:
A NAT is perfectly fine, since I want to know the different installation not the user particularly.

3nvi replied:
That's a default feature for every error logging service. I'll have to dig into Sentry to see if it allows me to configure it

3nvi replied:
Yeah turns out I can. I removed IP collection and will update the docs accordingly

austinbyers commented on docs/gitbook/privacy/error-reporting.md:
```diff
@@ -0,0 +1,15 @@
+Opting in or out of it is your choice, which you can always modify through your organization's
+**General Settings** page. If you opt in, the following information will be sent to the Panther team
+everytime there is a web application exception or crash:
+
```
Opting in or out of what?

```suggestion
If you opt in to error reporting, the following information will be sent to the Panther team every time there is a web application exception or crash:
```

Then, after the list,

```
You can always change your error reporting preference through the **General Settings** page
```

austinbyers commented on docs/gitbook/privacy/error-reporting.md:
```diff
@@ -0,0 +1,15 @@
+# Background
```
You will need to add the doc to the [summary](https://github.com/panther-labs/panther/blob/master/docs/gitbook/SUMMARY.md) for it to show up in the navigation bar

3nvi replied:
cool, thanks!

austinbyers commented on web/src/components/modals/analytics-consent-modal.tsx:
```diff
@@ -77,11 +77,12 @@ const AnalyticsConsentModal: React.FC = () => {
       disableEscapeKeyDown
       onClose={hideModal}
-      title="Help us make Panther better"
+      title="Help Improve Panther!"
```
👍

austinbyers commented on web/src/components/modals/analytics-consent-modal.tsx:
```diff
@@ -77,11 +77,12 @@ const AnalyticsConsentModal: React.FC = () => {
+      <Box width={600} px={100} pb={25}>
+        <Text size="large" color="grey300" mb={8}>
+          Opt-in to occasionally provide diagnostic information for improving reliability.
+          <b> All information is anonymized.</b>
```
```suggestion
Share application errors and crash logs with the Panther team to help us improve reliability
```

austinbyers commented on web/src/components/utils/modal-context.tsx:
```diff
@@ -34,6 +34,7 @@ export enum MODALS {
   DELETE_SOURCE = 'DELETE_SOURCE',
   DELETE_DESTINATION = 'DELETE_DESTINATION',
   NETWORK_ERROR = 'NETWORK_ERROR',
+  ANALYTICS_CONSENT = 'ANALYTICS_CONSENT',
```
Analytics consent and error logging conset are different or will be the same?

Right now, we only collect error logs, but could also include usage information in the future?

3nvi replied:
yeah, it's for the modals. I needed a name for both errors and actual activity logging. Analytics sounded like a good candidate, but thtat's something minor that can change easily
